DOC reorganize the build instructions (#15315)
scikit-learn/scikit-learn: doc/developers/advanced_installation.rst

.. _advanced-installation:

==================================================
Installing the development version of scikit-learn
==================================================

This section introduces how to install the **master branch** of scikit-learn.
This is required:

> - to test a feature that will be shipped in the next release (that is, a

Reviewer: ```suggestion
- to try a new feature that will be shipped in the next release (that is, a
```

  feature from a pull-request that was recently merged to the master branch);

- to check whether a bug you encountered has been fixed since the last
  release;

> - or to work on a contribution of your own.

Reviewer: make it the top entry since this is probably the most common case?

.. _advanced-installation:

==================================================
Installing the development version of scikit-learn
==================================================

This section introduces how to install the **master branch** of scikit-learn.
This is required:

- to work on a contribution (bug fix, new feature, code or documentation
  improvement);

- to try a new feature that will be shipped in the next release (that is, a

>   feature from a pull-request that was recently merged to the master branch);

Reviewer 2: We do have nightly wheels for master, so installing from source is probably not the easiest way to go. Maybe change to "build features from pull-requests"?

Author: This this the header for "Installing the development version of scikit-learn" and the first subsection is "Installing nightly builds".

Author: I pushed a new commit to split this list in the 2 sub-sections.

Reviewer 2: Indeed, didn't see the subsection as it was collapsed.. nevermind.

- to check whether a bug you encountered has been fixed since the last

>   release;

Reviewer 2: Same as a above

and upload wheel packages for the most recent Python version on a nightly
basis.

Installing a nightly build is the quickest way to:

- try a new feature that will be shipped in the next release (that is, a
  feature from a pull-request that was recently merged to the master branch);

- check whether a bug you encountered has been fixed since the last release.

::

  pip install --pre -f https://sklearn-nightly.scdn8.secure.raxcdn.com scikit-learn

.. _install_bleeding_edge:

Building from source
====================

Building from source is required to work on a contribution (bug fix, new
feature, code or documentation improvement).

.. _git_repo:

#. Use `Git <https://git-scm.com/>`_ to check out the latest source from the
   `scikit-learn repository <https://github.com/scikit-learn/scikit-learn>`_ on
   Github.::

        git clone git://github.com/scikit-learn/scikit-learn.git
        cd scikit-learn

   If you plan on submitting a pull-request, you should clone from your fork
   instead.

#. Install a compiler with OpenMP_ support for your platform. See intructions
   for :ref:`compiler_windows`, :ref:`compiler_macos`, :ref:`compiler_linux`

>    and :ref:`compiler_freebsd`.

Reviewer: Maybe we could add a note that for most linux distros this will just work out of the box?

And if we go for the new solution, maybe we can just say that openmp support isn't critical for most contributions?

Author: You still to install `build-essential` on debian / ubuntu for instance. gcc is not installed by default and first time contribution data scientist might never have installed any C/C++ compilers prior to contributing to scikit-learn at a scipy sprint for instance.

#. Optional (but recommended): create and activate a dedicated virtualenv_
   or `conda environment`_.

#. Install Cython_ and build the project with pip in :ref:`editable_mode`::

        pip install cython
        pip install --verbose --editable .

#. Check that the installed scikit-learn has a version number ending with
   `.dev0`::

    python -c "import sklearn; sklearn.show_versions()"

#. Please refer to the :ref:`developers_guide` and :ref:`pytest_tips` to run
   the tests on the module of your choice.

.. note::

    You will have to re-run the ``pip install --editable .`` command every time

>     the source code of a Cython file (ending in `.pyx` or `.pxd`).

Reviewer: ```suggestion
    the source code of a Cython file is updated (ending in `.pyx` or `.pxd`).
```

.. _install_bleeding_edge:

Building from source
====================

.. _git_repo:

#. Use `Git <https://git-scm.com/>`_ to check out the latest source from the
   `scikit-learn repository <https://github.com/scikit-learn/scikit-learn>`_ on

>    `Github <https://github.com>`::

Reviewer 2: There is probably no need to link to github.com? they will get there when clicking on the scikit-learn repo, and too many links can be confusing...

Author: Alright.

Building from source
=====================

.. _git_repo:

Use `Git <https://git-scm.com/>`_ to check out the latest source from the
`scikit-learn repository <https://github.com/scikit-learn/scikit-learn>`_ on
`Github <https://github.com>`_ and pip to build the project::

>     git clone git://github.com/scikit-learn/scikit-learn.git

Reviewer: We typically want contributors to clone from their fork.

Author: You can always add another remote later. Personally I prefer to use `ogrisel` as my fork and `origin` for the upstream repo.

Reviewer: Maybe just add a comment like "if you plan on submitting a pull-request, you should clone from your fork instead"

Author: Fixed.

    cd scikit-learn

    pip install cython

>     pip install --editable .

Reviewer: --verbose  ?

> Check that the installed scikit-learn with a version number ending with

Reviewer 3: ```suggestion
Check that the installed scikit-learn has a version number ending with
```

`.dev0`::

    python -c "import sklearn; sklearn.show_versions()"

Run the tests with ``pytest``::

    pip install pytest

>     pytest sklearn

Reviewer: I would not add this, we don't want contributors to think `pytest sklearn` is the proper way to test their changes.

IMO the small snippet above is enough to ensure the install went correctly

Author: What would you recommend to launch the tests instead? Just "pytest" without `sklearn`?

Reviewer: I would just not mention tests here.

If you think we should, maybe link to point 2 of https://scikit-learn.org/dev/developers/contributing.html#pull-request-checklist

Reviewer: Also I'll make a PR to update the testing section of the contributing guide because we recommend using `make` which is wrong

Reviewer: see https://github.com/scikit-learn/scikit-learn/pull/15377

Author: I have updated the PR to link to the developers guide and the pytest tips.

Please read below for more details and :ref:`platform system specific
instructions <platform_specific_instructions>` (e.g. in case compilation error
message).

.. note::

>     You will have to re-run the ``pip install --editable .`` command every

Reviewer: Maybe we can keep it simple and say 

> You will have to re-run the ``pip install --editable .`` command every time you modify a Cython file (ending in `.pyx` or `.pxd`) to re-compile the extensions.

The rest (branch switching) becomes implicit but hopefully obvious for someone who would ever touch Cython code.

    time the source code of a compiled extension is changed (for instance when
    switching branches or pulling changes from upstream). Compiled extensions
    are Cython files (ending in `.pyx` or `.pxd`). See :ref:`editable_mode`

>     for more details.

Reviewer: I'm not sure we need the ref here: this note is about building extensions, while the reference is about `--editable`. I would have expected the ref to be about extensions too.

Dependencies
------------

Runtime dependencies
~~~~~~~~~~~~~~~~~~~~

Scikit-learn requires the following dependencies both at build time and at
runtime:

- Python (>= 3.5),
- NumPy (>= 1.11),
- SciPy (>= 0.17),
- Joblib (>= 0.11).

Those dependencies are **automatically installed by pip and conda** when

> installing scikit-learn or building scikit-learn from source.

Reviewer 3: that is only true if you run `pip install -e .`
If you run `python setup.py build_ext -i` or equivalently `make in`, you need to install the dependencies first.

Reviewer: The text says `pip` so I guess that's fine

(not sure if your comment came first ^^)

Author: Yes we never mention any other way to build scikit-learn in this doc.

Reviewer 2: Maybe
```
Those dependencies are **automatically installed by pip* if they were missing when
building scikit-learn from source.
```
Saying that they will be installed with conda seems incorrect when building from source. They would be when conda installing scikit-learn, but since we are building from source section that part is probably more relevant? Or anyway reformulate this sentence to avoid confusing users on that point.

.. note::

   For running on PyPy, PyPy3-v5.10+, Numpy 1.14.0+, and scipy 1.1.0+
   are required. For PyPy, only installation instructions with pip apply.

Build dependencies
~~~~~~~~~~~~~~~~~~

Building Scikit-learn also requires:

- Cython >= 0.28.5
- A C/C++ compiler and a matching `OpenMP
  <https://en.wikipedia.org/wiki/OpenMP>`_ runtime library. See the
  :ref:`platform system specific instructions
  <platform_specific_instructions>` for more details.

.. note::

   It is possible to build scikit-learn without OpenMP support by setting the
   ``SKLEARN_NO_OPENMP`` environment variable (before cythonization). This is
   not recommended since it will force some estimators to run in sequential
   mode.

Since version 0.21, scikit-learn automatically detects and use the
linear algebrea library used by SciPy **at runtime**. Scikit-learn has
therefore no build dependency on BLAS/LAPACK implementations such as OpenBlas,

> Atlas or MKL.

Reviewer 3: OpenBlas, Atlas, BLIS or MKL. :)

Test dependencies
~~~~~~~~~~~~~~~~~

Running tests requires:

> .. |PytestMinVersion| replace:: 3.3.0

Reviewer: Actually there are bugs in `pytest` as low as 3.8.1 

@thomasjpfan recently updated the CI to 4.6.2 I think

Author: Fixed.

to build scikit-learn Cython extensions for each supported platform.

Windows
-------

First, install `Build Tools for Visual Studio 2019
<https://visualstudio.microsoft.com/downloads/>`_.

.. warning::

    You DO NOT need to install Visual Studio 2019. You only need the "Build
    Tools for Visual Studio 2019", under "All downloads" -> "Tools for Visual
    Studio 2019".

Secondly, find out if you are running 64-bit or 32-bit Python. The building
command depends on the architecture of the Python interpreter. You can check
the architecture by running the following in ``cmd`` or ``powershell``
console::

    python -c "import struct; print(struct.calcsize('P') * 8)"

For 64-bit Python, configure the build environment with::

    SET DISTUTILS_USE_SDK=1
    "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Auxiliary\Build\vcvarsall.bat" x64

Replace ``x64`` by ``x86`` to build for 32-bit Python.

Please be aware that the path above might be different from user to user. The
aim is to point to the "vcvarsall.bat" file that will set the necessary
environment variables in the current command prompt.

Finally, build scikit-learn from this command prompt::

>     pip install --editable .

Reviewer 4: ```suggestion
    pip install --verbose --editable .
```

macOS
-----

The default C compiler on macOS, Apple clang (confusingly aliased as
`/usr/bin/gcc`), does not directly support OpenMP. We present two alternatives
to enable OpenMP support:

- either install `conda-forge::compilers` with conda;

- or install `libomp` with Homebrew to extend the default Apple clang compiler.

macOS compilers from conda-forge
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you use the conda package manager, you can install the ``compilers``
meta-package from the conda-forge channel, which provides OpenMP-enabled C/C++
compilers based on the llvm toolchain.

It is recommended to use a dedicated conda environment to build scikit-learn
from source::

    conda create -n sklearn-dev python numpy scipy cython joblib pytest \
        conda-forge::compilers conda-forge::llvm-openmp
    conda activate sklearn-dev

>     make clean

Reviewer 4: Do we want to advise users to `make clean` in the other instructions?

Reviewer: corollary: why is `make clean` needed here and not elsewhere?

Reviewer 3: because on macOS you may have build sklearn once without OpenMP support, then came to the doc to see how to enable OpenMP. In that case you need to clean first.

Author: I don't think it's useful for the other cases.

    echo $CXXFLAGS
    echo $LDFLAGS

They point to files and folders from your ``sklearn-dev`` conda environment
(in particular in the bin/, include/ and lib/ subfolders).

The compiled extensions should be built with the clang and clang++ compilers

> installed by conda with the ``-fopenmp`` command line flag.

Reviewer: In the log, you should see the compiled extension being built with...

    export CXXFLAGS="$CXXFLAGS -I/usr/local/opt/libomp/include"
    export LDFLAGS="$LDFLAGS -Wl,-rpath,/usr/local/opt/libomp/lib -L/usr/local/opt/libomp/lib -lomp"

Finally, build scikit-learn in verbose mode (to check for the presence of the
``-fopenmp`` flag in the compiler commands)::

    make clean
    pip install --verbose --editable .

Linux
-----

Linux compilers from the system
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Installing scikit-learn from source without using conda requires you to have
installed the scikit-learn Python development headers and a working C/C++
compiler with OpenMP support (typically the GCC toolchain).

> Install build dependencies for Debian-based operating systems, which include

Reviewer: ```suggestion
Install build dependencies for Debian-based operating systems, e.g.
```

(not sure whether "include" takes an s here?)

Author: "operating systems" is plural so no "s". But fair enough.

Ubuntu::

    sudo apt-get install build-essential python3-dev python3-pip

then proceed as usual::

    pip3 install cython

>     pip3 install --editable .

Reviewer 4: ```suggestion
    pip3 install --verbose --editable .
```

When precompiled wheels of the runtime dependencies are not avalaible for your
architecture (e.g. ARM), you can install the system versions::

    sudo apt-get install cython3 python3-numpy python3-scipy

On Red Hat and clones (e.g. CentOS), install the dependencies using::

    sudo yum -y install gcc gcc-c++ python-devel numpy scipy

Linux compilers from conda-forge
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Alternatively, install a recent version of the GNU C Compiler toolchain (GCC)
in the user folder using conda::

    conda create -n sklearn-dev numpy scipy joblib cython conda-forge::compilers
    conda activate sklearn-dev

>     pip install --editable .

Reviewer 4: ```suggestion
    pip install --verbose --editable .
```

FreeBSD
-------

The clang compiler included in FreeBSD 12.0 and 11.2 base systems does not
include OpenMP support. You need to install the `openmp` library from packages
(or ports)::

    sudo pkg install openmp

This will install header files in ``/usr/local/include`` and libs in
``/usr/local/lib``. Since these directories are not searched by default, you
can set the environment variables to these locations::

    export CFLAGS="$CFLAGS -I/usr/local/include"
    export CXXFLAGS="$CXXFLAGS -I/usr/local/include"
    export LDFLAGS="$LDFLAGS -Wl,-rpath,/usr/local/lib -L/usr/local/lib -lomp"

Finally, build the package using the standard command::

>     pip install --editable .

Reviewer 4: ```suggestion
    pip install --verbose --editable .
```
